transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,813 @@
+        }
+    }
+
+    pub(crate) fn handle_event(
+        &mut self,
+        event: quinn_proto::ConnectionEvent,
+    ) -> Option<quinn_proto::EndpointEvent> {
+        self.connection.handle_event(event);
+        self.connection.poll_endpoint_events()
+    }
+
+    /// Check if this connection is drained.
+    pub(crate) fn is_drained(&self) -> bool {
+        self.connection.is_drained()
+    }
+
+    /// Accept an incoming stream, if possible.
+    pub(crate) fn accept(&mut self, cx: &mut Context<'_>) -> Poll<quinn_proto::StreamId> {
+        match self.connection.accept(quinn_proto::Dir::Bi) {
+            None => {
+                self.set_waker(cx);
+                Poll::Pending
+            }
+            Some(e) => Poll::Ready(e),
+        }
+    }
+
+    /// Open an outgoing stream if it is possible to do so.
+    pub(crate) fn open(&mut self) -> Option<quinn_proto::StreamId> {
+        self.connection.open(quinn_proto::Dir::Bi)
+    }
+
+    /// Write to a stream.
+    pub(crate) fn write(
+        &mut self,
+        id: quinn_proto::StreamId,
+        buffer: &[u8],
+    ) -> Result<usize, quinn_proto::WriteError> {
+        self.connection.write(id, buffer)
+    }
+
+    /// Shut down the writing side of a stream.
+    pub(crate) fn finish(
+        &mut self,
+        id: quinn_proto::StreamId,
+    ) -> Result<(), quinn_proto::FinishError> {
+        self.connection.finish(id)
+    }
+
+    /// Get application-facing events
+    pub(crate) fn poll(&mut self) -> Option<quinn_proto::Event> {
+        self.connection.poll()
+    }
+
+    /// Get the side of the stream
+    pub(crate) fn side(&self) -> quinn_proto::Side {
+        self.connection.side()
+    }
+
+    /// Read from a stream
+    pub(crate) fn read(
+        &mut self,
+        id: quinn_proto::StreamId,
+        buf: &mut [u8],
+    ) -> Result<usize, quinn_proto::ReadError> {
+        self.connection.read(id, buf).map(|size| size.unwrap_or(0))
+    }
+
+    /// Close the stream, if it is not closed already
+    pub(crate) fn close(&mut self, status: u32) {
+        if self.connection.is_closed() {
+            return;
+        }
+        self.connection.close(
+            std::time::Instant::now(),
+            quinn_proto::VarInt::from_u32(status),
+            Default::default(),
+        )
+    }
+}
+
+mod channel_ref {
+    use super::{Arc, EndpointData, EndpointMessage::Dummy, Sender};
+    #[derive(Debug, Clone)]
+    pub(super) struct Channel(Sender);
+
+    impl Channel {
+        pub(super) fn new(s: Sender) -> Self {
+            Self(s)
+        }
+    }
+
+    impl std::ops::Deref for Channel {
+        type Target = Sender;
+        fn deref(&self) -> &Sender {
+            &self.0
+        }
+    }
+
+    impl std::ops::DerefMut for Channel {
+        fn deref_mut(&mut self) -> &mut Sender {
+            &mut self.0
+        }
+    }
+
+    impl Drop for Channel {
+        fn drop(&mut self) {
+            let _ = self.0.start_send(Dummy);
```

Comment:
Can you add a note why one needs to send a `Dummy` before dropping?

Reply:
It’s to wake up the endpoint and avoid deadlocks, but indeed there should be a comment.